src/main/java/com/friendly_machines/intellij/plugins/ideanative2debugger/impl/GdbMiFilter.java

```diff
@@ -79,45 +79,47 @@ private static byte[] maybeEscape(byte[] text) {
             }
         }
         if (escaping_needed) {
-            return makeCString(text);
+            makeCString(text, out);
+            return;
         } else {
-            return text;
+            out.write(text);
+            return;
         }
     }
 
-    public GdbMiStateResponse gdbSend(String operation, String[] options, String[] parameters) {
+    public GdbMiStateResponse gdbSend(String operation, Collection<String> options, Collection<String> parameters) throws IOException, InterruptedException {
 //        println("gdbSend " + operation);
-        try {
-            ++requestId;
-            myChildIn.write(Integer.toString(requestId).getBytes(StandardCharsets.UTF_8));
-            myChildIn.write(operation.getBytes(StandardCharsets.UTF_8));
-            for (String option : options) {
-                myChildIn.write(" ".getBytes(StandardCharsets.UTF_8));
-                myChildIn.write(maybeEscape(option.getBytes(StandardCharsets.UTF_8)));
-            }
-            if (parameters.length > 0) {
-                myChildIn.write(" --".getBytes(StandardCharsets.UTF_8));
-                for (String parameter : parameters) {
-                    myChildIn.write(" ".getBytes(StandardCharsets.UTF_8));
-                    myChildIn.write(maybeEscape(parameter.getBytes(StandardCharsets.UTF_8)));
-                }
+        ++requestId;
+        myChildIn.print(Integer.toString(requestId));
+        myChildIn.print(operation);
+        for (String option : options) {
+            myChildIn.print(" ");
+            maybeEscape(option.getBytes(StandardCharsets.UTF_8), myChildIn);
+        }
+        if (!parameters.isEmpty()) {
+            myChildIn.print(" --");
+            for (var parameterStr : parameters) {
+                myChildIn.print(" ");
+                maybeEscape(parameterStr.getBytes(StandardCharsets.UTF_8), myChildIn);
             }
-            myChildIn.write("\r\n".getBytes(StandardCharsets.UTF_8));
-            myChildIn.flush();
-            return readResponse();
-        } catch (IOException | InterruptedException e) {
-            e.printStackTrace();
-            throw new RuntimeException(e);
         }
+        myChildIn.print("\r\n");
+        myChildIn.flush();
+        return readResponse();
     }
 
-    public Map<String, Object> gdbCall(String operation, String[] options, String[] parameters) throws GdbMiOperationException {
-        GdbMiStateResponse response = gdbSend(operation, options, parameters);
+    public Map<String, ?> gdbCall(String operation, Collection<String> options, Collection<String> parameters) throws GdbMiOperationException, IOException, InterruptedException {
+        var response = gdbSend(operation, options, parameters);
         assert response.getMode() == '^';
```

Comment:
Better to also throw an exception instead of asserting ? Just GdbMiOperationException maybe.